File: 0330-patching-array/0330-patching-array.java
class Solution {
    public int minPatches(int[] nums, int n) {
        // maxReachableSum represents the maximum sum we can form using current elements
        // We can form all sums in range [1, maxReachableSum - 1]
        long maxReachableSum = 1;
      
        // Count of patches (new numbers) we need to add
        int patchCount = 0;
      
        // Index to traverse the nums array
        int index = 0;
      
        // Continue until we can form all sums up to n
        while (maxReachableSum <= n) {
            // Check if we can use the current element from nums array
            if (index < nums.length && nums[index] <= maxReachableSum) {
                // If current element is within our reachable range,
                // we can extend our range by adding this element
                maxReachableSum += nums[index];
                index++;
            } else {
                // If current element is too large or we've exhausted the array,
                // we need to add a patch (the optimal patch is maxReachableSum itself)
                patchCount++;
              
                // Adding maxReachableSum as a patch doubles our reachable range
                // This is equivalent to: maxReachableSum = maxReachableSum * 2
                maxReachableSum <<= 1;
            }
        }
      
        return patchCount;
    }
}

File: 0330-patching-array/0330-patching-array.ts
/**
 * Calculates the minimum number of patches required to form every integer in range [1, n]
 * @param nums - Sorted array of positive integers
 * @param n - Target upper bound of the range
 * @returns Minimum number of patches needed
 */
function minPatches(nums: number[], n: number): number {
    // The smallest number that cannot be formed with current elements
    let smallestMissing: number = 1;
  
    // Count of patches added
    let patchCount: number = 0;
  
    // Index for traversing the nums array
    let index: number = 0;
  
    // Continue until we can form all numbers up to n
    while (smallestMissing <= n) {
        if (index < nums.length && nums[index] <= smallestMissing) {
            // If current number can help extend our range, use it
            // We can now form all numbers up to (smallestMissing + nums[index] - 1)
            smallestMissing += nums[index];
            index++;
        } else {
            // Need to patch: add smallestMissing itself
            // This doubles our reachable range
            patchCount++;
            smallestMissing *= 2;
        }
    }
  
    return patchCount;
}
